aws/aws-toolkit-jetbrains
# Enable backend gutter marks for lambda functions

Add ability to run/debug and create new lambda from gutter mark.

## Types of changes

## Description
- Add AWS Rider backend C# logic
- Update gradle to compile AWS Rider backend logic and copy it into plugin output
- Add Rider Protocol extensions to handle backend events
- Update gradle to generate protocol models on Kotlin and C# sides
- Add fake backend icons for actions in Lambda gutter mark context
- Add Icons patcher on frontend to update fake backend icons with frontend icons
- Pass credentials when opening run lambda run configuration
- Renable the Configuration type instance to separate from configuration type
- Add frontend tests to verify gutter marks

## Motivation and Context
Rider does not work with Idea PSI and do not have information about csharp structure on a frontend. It is processed in ReSharper process (backend) instead. Gutter mark to run/debug/create lambda function should be calculated on a backend. This change set up the backend and protocol in plugin and enable gutter marks from backend. selecting an action on a backend fire an event on a frontend through protocol, so, it can be properly handled whether by calling a run configuration to run lambda, or call an action to create a new lambda.

## Related Issue(s)
None

## Testing
```
JetBrains Rider 2019.2 EAP
Runtime version: 1.8.0_212-release-1566-b8 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.5
GC: PS Scavenge, PS MarkSweep
Memory: 480M
Cores: 8
```
Verified new gutter mark: 
- Run/debug/create actions
- Verify lambda detection for cases described in https://docs.aws.amazon.com/lambda/latest/dg/dotnet-programming-model-handler-types.html

## Screenshots (if appropriate)
<img width="652" alt="LambdaGutterMark" src="https://user-images.githubusercontent.com/6064345/62258804-97c31180-b414-11e9-8984-29a25ac778a7.png">

## Checklist

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.
